filestore/filepath.go

```diff
@@ -0,0 +1,399 @@
+	//case DB:
+	//	return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	case HDFS:
+		return nil, fmt.Errorf("currently unsupported file store type '%s'", storeType)
+	default:
+		return nil, fmt.Errorf("unknown store type '%s'", storeType)
+	}
+}
+
+type FilePath struct {
+	scheme  string
+	bucket  string
+	key     string
+	isDir   bool
+	isValid bool
+}
+
+func (fp *FilePath) SetScheme(scheme string) {
+	fp.scheme = scheme
+}
+
+func (fp *FilePath) Scheme() string {
+	return fp.scheme
+}
+
+func (fp *FilePath) SetBucket(bucket string) {
+	fp.bucket = bucket
+}
+
+func (fp *FilePath) Bucket() string {
+	return fp.bucket
+}
+
+func (fp *FilePath) SetKey(key string) {
+	fp.key = key
+}
+
+func (fp *FilePath) Key() string {
+	return fp.key
+}
+
+func (fp *FilePath) KeyPrefix() string {
+	return filepath.Dir(fp.key)
+}
+
+func (fp *FilePath) Ext() FileType {
+	ext := filepath.Ext(fp.key)
+	// filepath.Ext returns the extension with the "." prefix, so we need to trim it
+	// to match our FileType type.
+	return FileType(strings.TrimPrefix(ext, "."))
+}
+
+func (fp *FilePath) PathWithBucket() string {
+	return fp.key
+}
+
+func (fp *FilePath) SetIsDir(isDir bool) {
+	fp.isDir = isDir
+}
+
+func (fp *FilePath) IsDir() bool {
+	return fp.isDir
+}
+
+func (fp *FilePath) ParseFilePath(fullPath string) error {
+	err := fp.parsePath(fullPath)
+	if err != nil {
+		return fmt.Errorf("file: %v", err)
+	}
+	fp.isDir = false
+	return nil
+}
+
+func (fp *FilePath) ParseDirPath(fullPath string) error {
+	err := fp.parsePath(fullPath)
+	if err != nil {
+		return fmt.Errorf("dir: %v", err)
+	}
+	// TODO: consider removing the final piece of the path if it's a file
+	fp.isDir = true
+	return nil
+}
+
+func (fp *FilePath) checkSchemes(scheme string) error {
+	for _, s := range ValidSchemes {
+		if s == scheme {
+			return nil
+		}
+	}
+	return fmt.Errorf("invalid scheme '%s', must be one of %v", scheme, ValidSchemes)
+}
+
+func (fp *FilePath) parsePath(fullPath string) error {
+	// Parse the URI into a url.URL object.
+	u, err := url.Parse(fullPath)
+	if err != nil {
+		return fmt.Errorf("could not parse fullpath '%s': %v", fullPath, err)
+	}
+
+	// Extract the bucket and path components from the URI.
+	bucket := u.Host
+	path := strings.TrimPrefix(u.Path, "/")
+
+	err = fp.checkSchemes(u.Scheme)
+	if err != nil {
+		return err
+	} else {
+		fp.scheme = u.Scheme
+	}
+
+	fp.bucket = bucket
+	fp.key = path
+	return nil
+}
+
+func (fp *FilePath) IsValid() bool {
+	return fp.isValid
+}
+
+func (fp *FilePath) Validate() error {
+	return fmt.Errorf("not implemented")
+}
+
+type S3Filepath struct {
+	FilePath
+}
+
+func (s3 *S3Filepath) Validate() error {
+	if s3.scheme != "s3://" && s3.scheme != "s3a://" {
+		return fmt.Errorf("invalid scheme '%s', must be 's3:// or 's3a://'", s3.scheme)
+	}
+	if s3.bucket == "" {
+		return fmt.Errorf("bucket cannot be empty")
+	} else {
+		s3.bucket = strings.Trim(s3.bucket, "/")
+	}
+	if s3.key == "" {
```

Comment:
would we have cases where `key == "/"`?

Reply:
We shouldn't but, I could add that to the conditional check to be sure.